src/core/dense_set.cc

```diff
@@ -0,0 +1,558 @@
+  ChainIterator prev;
+  for (auto it = ChainBegin(bid); it != ChainEnd(bid); ++it) {
+    if (Equal(*it, ptr)) {
+      return {it, prev};
+    }
+
+    prev = it;
+  }
+
+  // not in the Set
+  return {ChainEnd(bid), ChainEnd(bid)};
+}
+
+// Same idea as FindAround but provide the const guarantee
+bool DenseSet::ContainsInternal(const void* ptr) const {
+  uint32_t bid = BucketId(ptr);
+  if (bid && Equal(&entries_[bid - 1], ptr)) {
+    return true;
+  }
+
+  if (bid + 1 < entries_.size() && Equal(&entries_[bid + 1], ptr)) {
+    return true;
+  }
+
+  for (auto it = ChainConstBegin(bid); it != ChainConstEnd(bid); ++it) {
+    if (Equal(*it, ptr)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void* DenseSet::EraseInternal(void* ptr) {
+  uint32_t bid = BucketId(ptr);
+  auto [found, prev] = FindAround(ptr, bid);
+
+  if (found == nullptr) {
+    return nullptr;
+  }
+
+  if (found->IsLink()) {
+    --num_chain_entries_;
+  } else {
+    DCHECK(found->IsObject());
+    --num_used_buckets_;
+  }
+
+  obj_malloc_used_ -= ObjectAllocSize(ptr);
+  void* ret;
+  (void)UnlinkAndFree(found, prev, &ret);
+
+  --size_;
+  return ret;
+}
+
+void* DenseSet::PopInternal() {
+  std::pmr::vector<DenseSet::DensePtr>::iterator bucket_iter = entries_.begin();
+
+  // find the first non-empty chain
+  while (bucket_iter != entries_.end() && bucket_iter->IsEmpty()) {
+    ++bucket_iter;
+  }
+
+  // empty set
+  if (bucket_iter == entries_.end()) {
+    return nullptr;
+  }
+
+  if (bucket_iter->IsLink()) {
+    --num_chain_entries_;
+  } else {
+    DCHECK(bucket_iter->IsObject());
+    --num_used_buckets_;
+  }
+
+  // unlink the first node in the first non-empty chain
+  obj_malloc_used_ -= ObjectAllocSize(bucket_iter->GetObject());
+  void* ret;
+  (void)UnlinkAndFree(&*bucket_iter, nullptr, &ret);
+
+  --size_;
+  return ret;
+}
+
+/**
+ * stable scanning api. has the same guarantees as redis scan command.
+ * we avoid doing bit-reverse by using a different function to derive a bucket id
+ * from hash values. By using msb part of hash we make it "stable" with respect to
+ * rehashes. For example, with table log size 4 (size 16), entries in bucket id
+ * 1110 come from hashes 1110XXXXX.... When a table grows to log size 5,
+ * these entries can move either to 11100 or 11101. So if we traversed with our cursor
+ * range [0000-1110], it's guaranteed that in grown table we do not need to cover again
+ * [00000-11100]. Similarly with shrinkage, if a table is shrinked to log size 3,
+ * keys from 1110 and 1111 will move to bucket 111. Again, it's guaranteed that we
+ * covered the range [000-111] (all keys in that case).
+ * Returns: next cursor or 0 if reached the end of scan.
+ * cursor = 0 - initiates a new scan.
+ *
+ * Therefore for any given Scan cursor x which was run when capacity_log_ was m, if the
+ * next Scan is run when capacity_log_ is n the current bucket to check is x << (n - m)
```

Comment:
I am not sure I understand this statement and the optimization that follows it.
Given the cursor we always take high capacity_log_ bits in the cursor whatever they are. We shift the cursor **right** by `(32-capacity_log_)` to bring those bits to us.

Reply:
Essentially the optimization I was trying to describe is if the last cursor is `101` with capacity log 4, if during the next scan the capacity log has grown to 5 the cursor can jump to `1010` instead of staying at `101`

Reply:
It should happen automatically in the code we had before.  This is why we pad to the left of uint32_t cursor:

```py
0b101 << (32 - 4)   #  1342177280
bin(1342177280)    # 0b1010000000000000000000000000000
1342177280 >> (32 - 5)   # 0b1010
```